### PR TITLE
Added Evaluation of Literal Types

### PR DESCRIPTION
**Pull request by ggetz**
Simply check for literal type and evaluate to the literal's value.

I added specs to check for evaluation of literal:

* null
* boolean
* number
* string
* Color
  * color constructor that takes a css format string - color('white'), color('#ffffff')
  * rgb constructor - rgb(255, 255, 255)

I also added a structure `Node` for constructing our AST. It's currently incomplete, it doesn't have any children nodes. I define the type of expression with an integer, (LITERAL = 0, UNARY = 1, and BINARY = 2), and the value.

### CODE REVIEW

**Comment by pjcozzi** [comment]
Replaces https://github.com/ggetz/cesium/pull/3.

File: Source/Scene/Expression.js [hunk 1]
```diff
@@ -1,26 +1,75 @@
...
-       '../ThirdParty/jsep',
+       '../ThirdParty/jsep'
     ], function(
+        Color,
+        defined,
         defineProperties,
         jsep) {
     "use strict";
 
+    var LITERAL = 0;
```
Comment (pjcozzi) [inline Source/Scene/Expression.js]: Put these in an "enum" (in this file, for now) like https://github.com/AnalyticalGraphicsInc/cesium/blob/3d-tiles/Source/Scene/TileState.js.

File: Source/Scene/Expression.js [hunk 2]
```diff
@@ -1,26 +1,75 @@
...
+
     /**
      * DOC_TBA
      */
     function Expression(styleEngine, expression) {
         this._styleEngine = styleEngine;
 
-        console.log(jsep(expression));
+        var ast = jsep(expression);
+        this._runtimeAst = undefined;
```
Comment (pjcozzi) [inline Source/Scene/Expression.js]: This line isn't needed, `_runtimeAst` is declared and initialized below.

File: Source/Scene/Expression.js [hunk 3]
```diff
@@ -1,26 +1,75 @@
...
+        }
+
+        return node;
+    }
+
     Expression.prototype.evaluate = function(feature) {
-        return true;
+        var runtimeAst = this._runtimeAst;
+
+        if (runtimeAst._type === LITERAL) {
```
Comment (pjcozzi) [inline Source/Scene/Expression.js]: Instead of doing this check here, I suspect we are going to want to do this check in `createRuntimeAst` and assign a particular function like we do now in, for example, `BooleanExpression`.  Do you see what I mean?

File: Source/Scene/Expression.js [hunk 4]
```diff
@@ -1,26 +1,75 @@
...
+        this._value = value;
+    }
+
+    function createRuntimeAst(ast) {
+        var node;
+
+        if (ast.type === 'Literal') {
+            node = new Node(LITERAL, ast.value);
+        } else if (ast.type === 'CallExpression') {
+            var call = ast.callee.name;
```
Comment (pjcozzi) [inline Source/Scene/Expression.js]: It can be a separate PR, but there are a lot of syntax errors we should catch here.

File: Source/Scene/Expression.js [hunk 5]
```diff
@@ -1,26 +1,75 @@
...
+            var args = ast.arguments;
+            var val;
+            if (call === 'color') {
+                 val = Color.fromCssColorString(args[0].value);
```
Comment (pjcozzi) [inline Source/Scene/Expression.js]: Whitespace.

**Comment by pjcozzi** [comment]
Why not include `hsl` in this PR to round out color literals?

**Comment by pjcozzi** [comment]
That's all my comments.  This is heading in the right direction.

File: Source/Scene/Expression.js (hunk 3 above)
Comment (ggetz) [inline Source/Scene/Expression.js]: OK, I see.

**Comment by ggetz** [comment]
Updated. I added an evaluation function to the `Node` structure, and assign it when creating the runtime AST.

File: Source/Scene/Expression.js [hunk 6]
```diff
@@ -13,14 +19,58 @@ define([
...
+
+        return node;
+    }
+
     Expression.prototype.evaluate = function(feature) {
-        return true;
+        return this._runtimeAst.evaluate(feature);
+    };
+
+    Expression.prototype._evaluateLiteral = function(feature) {
```
Comment (pjcozzi) [inline Source/Scene/Expression.js]: This should be part of `Node`, not `Expression`.

File: Source/Scene/Expression.js [hunk 7]
```diff
@@ -13,14 +19,58 @@ define([
...
+
+        this._runtimeAst = createRuntimeAst(this, ast);
+        console.log(this._runtimeAst);
     }
 
     defineProperties(Expression.prototype, {
     });
 
+    function Node(type, value, evaluate) {
+        this._type = type;
```
Comment (pjcozzi) [inline Source/Scene/Expression.js]: Why explicitly store `type` now?  If we need it later, e.g., to expose as part of the API or for type conversion then let's add it then.

File: Source/Scene/ExpressionType.js [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+define([
+    '../Core/freezeObject'
+], function(
+    freezeObject) {
+    'use strict';
+
+    /**
+     * @private
+     */
+    var ExpressionType = {
```
Comment (pjcozzi) [inline Source/Scene/ExpressionType.js]: We need to be rigorous with terminology.  This is not an expression type.  This is the type of a node in an AST.

Comment (ggetz) [inline Source/Scene/ExpressionType.js]: Should I still include this file if we are not storing type for now?

Comment (pjcozzi) [inline Source/Scene/ExpressionType.js]: Ah, right, I wouldn't bother until it is needed.

**Comment by ggetz** [comment]
OK, updated.

**Comment by pjcozzi** [comment]
Looks good!
